rmf_site_format/src/scenario.rs

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2024 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+use crate::*;
+#[cfg(feature = "bevy")]
+use bevy::prelude::{Bundle, Component, Reflect, ReflectComponent};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "bevy", derive(Component, Reflect))]
+#[cfg_attr(feature = "bevy", reflect(Component))]
+pub struct InstanceMarker;
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy, Default, PartialEq, Eq)]
+#[cfg_attr(feature = "bevy", derive(Component, Reflect))]
+#[cfg_attr(feature = "bevy", reflect(Component))]
+pub struct ScenarioMarker;
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "bevy", derive(Component))]
+pub struct Scenario<T: RefTrait> {
+    pub parent_scenario: Affiliation<T>,
+    pub added_instances: Vec<(T, Pose)>,
+    pub removed_instances: Vec<T>,
+    pub moved_instances: Vec<(T, Pose)>,
+}
+
+impl<T: RefTrait> Scenario<T> {
+    pub fn from_parent(parent: T) -> Scenario<T> {
+        Scenario {
+            parent_scenario: Affiliation(Some(parent)),
+            added_instances: Vec::new(),
+            removed_instances: Vec::new(),
+            moved_instances: Vec::new(),
+        }
+    }
+}
+
+// Create a root scenario without parent
+impl<T: RefTrait> Default for Scenario<T> {
+    fn default() -> Self {
+        Self {
+            parent_scenario: Affiliation::default(),
+            added_instances: Vec::new(),
+            removed_instances: Vec::new(),
+            moved_instances: Vec::new(),
+        }
+    }
+}
+
+impl<T: RefTrait> Scenario<T> {
+    pub fn convert<U: RefTrait>(&self, id_map: &HashMap<T, U>) -> Result<Scenario<U>, T> {
```

Comment:
I wonder if [this question](https://github.com/open-rmf/rmf_site/pull/242#discussion_r1898459969) was asking about the `.expect`s in this function?

If so, here's what I would recommend:

```rust
        Ok(Scenario {
            parent_scenario: self.parent_scenario.convert(id_map)?,
            added_instances: self
                .added_instances
                .clone()
                .into_iter()
                .map(|(id, pose)| {
                    let converted_id = id_map.get(&id).cloned().ok_or(id)?;
                    Ok((converted_id, pose))
                })
                .collect::<Result<_, _>>()?,
            removed_instances: self
                .removed_instances
                .clone()
                .into_iter()
                .map(|id| id_map.get(&id).cloned().ok_or(id))
                .collect::<Result<_, _>>()?,
            moved_instances: self
                .moved_instances
                .clone()
                .into_iter()
                .map(|(id, pose)| {
                    let converted_id = id_map.get(&id).cloned().ok_or(id)?;
                    Ok((converted_id, pose))
                })
                .collect::<Result<_, _>>()?,
        })
```

Reply:
Yes it is! Thanks for sharing.

Reply:
[812c183](https://github.com/open-rmf/rmf_site/pull/242/commits/812c183d415550ffe9152cebc083bb1dbd4b570b)